File: src/main/java/ninja/hudy/infosmog/facade/CountyFacade.java
package ninja.hudy.infosmog.facade;

import ninja.hudy.infosmog.model.CountyRegion;
import ninja.hudy.infosmog.model.Measurement;
import ninja.hudy.infosmog.model.RegionType;
import ninja.hudy.infosmog.name.DefaultName;
import ninja.hudy.infosmog.repository.MeasurementRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.Set;
import java.util.stream.Stream;

@Component
public class CountyFacade {

    @Autowired
    private MeasurementRepository measurementRepository;


    public Stream<CountyRegion> getCountiesWithDataAfter(Long timestamp) {
        Stream<CountyRegion> result = findMeasurements(timestamp).stream()
                .filter(measurement -> "PM10".equalsIgnoreCase(measurement.getType().getName()) || "NO2".equalsIgnoreCase(measurement.getType().getName()))
                .map(measurement -> measurement.getStation())
                .distinct()
                .map(measurement -> measurement.getCity())
                .distinct()
                .map(region -> region.getSuperRegion())
                .distinct()
                .filter(region -> RegionType.Values.COUNTY.equals(region.getType()) && !DefaultName.UNDEFINED_NAME.getName().equals(region.getName()))
                .map(region -> (CountyRegion) region);

        return result;
    }


    private Set<Measurement> findMeasurements(long timestamp) {
        return measurementRepository.findByTimestampAfter(timestamp);
    }
}

File: src/main/kotlin/pl/org/smok/infosmog/facade/CountyFacade.kt
package pl.org.smok.infosmog.facade

import pl.org.smok.infosmog.name.DefaultName
import pl.org.smok.infosmog.repository.MeasurementRepository
import pl.org.smok.infosmog.model.CountyRegion
import pl.org.smok.infosmog.model.RegionType
import pl.org.smok.infosmog.model.Measurement
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class CountyFacade {

    @Autowired
    lateinit var measurementRepository: MeasurementRepository;


    fun getCountiesWithDataAfter(timestamp: Long): List<CountyRegion> {
        val result = findMeasurements(timestamp)
                .filter {
                    "PM10".equals(it.type?.name, ignoreCase = true) || "NO2".equals(it.type?.name, ignoreCase = true)
                }
                .map { it.station }
                .distinct()
                .mapNotNull { it?.city }
                .distinct()
                .mapNotNull { it.superRegion }
                .distinct()
                .filter { RegionType.Values.COUNTY == it.type && DefaultName.UNDEFINED_NAME.name != it.name }
                .map { region -> region as CountyRegion }

        return result
    }


    private fun findMeasurements(timestamp: Long): Set<Measurement> {
        return measurementRepository.findByTimestampAfter(timestamp)
    }
}
